### ISSUE TITLE
Feature Request: Run loops should be parallizable

### ISSUE THREAD

**Issue by johngro** [domokit/mojo#679]
Currently, my understanding is that the mojo RunLoops are serviced by a single thread, and are internally thread hostile.  I would like to be able to create (or have created for me when my application starts) a run loop which can be serviced by a pool of threads.  Conceptually, whenever one or more MessageLoops associated with the RunLoop has a pending message and there is one or more idle threads in the pool, a thread from the pool should be assigned one of the message loops with pending work and dispatch the work.

Multiple threads should never service a messageloop (or more generically, handle) concurrently.  The messageloop should behave like a serialization context.

Note: I don't mean to imply that all RunLoop should be able to be serviced by multiple threads.  I would just like to be able to create an instance of something which exposes a run loop interface (see issue #645) which is backed by a pool of threads instead of just one.

@j9brown @chrismckillop @viettrungluu-cr @dalesat

**Comment by jamesr** [comment]
To clarify, you're specifically talking about in the C++ bindings and library layer here, right?  Not all languages act in this way (Go doesn't really have a message loop or run loop equivalent, some of the languages we bind to are single-threaded, etc).

**Comment by johngro** [comment]
@jamesr 

Correct.  While I think it would be great if this could be realized in other languages, my only real concern is C++ (because I'm selfish like that)
